In [3]:
def gini_index_single(a, b):
  single_gini = 1 - ((a/(a+b))**2) - ((b/(a+b))**2)
  return round(single_gini, 2)

In [4]:
print(gini_index_single(105, 39))

0.39


In [5]:
def gini_index(a, b, c, d):
    left = gini_index_single(a, b)
    right = gini_index_single(c, d)
    gini_index = left * ((a + b)/ (a + b + c + d)) + right * ((c + d)/(a + b + c + d))
    return round(gini_index, 3)

In [6]:
print(gini_index(105, 39, 34, 125)) # Chest Pain
print(gini_index(37, 127, 100, 33)) # Good Blood Circulation
print(gini_index(92, 31, 45, 129))  # Blocked Arteries

0.364
0.359
0.38


# The output results of the above three attributes are 0.364, 0.36, 0.381
# The smaller the gini index, the higher the purity, so choose the second attribute for the first node

In [7]:
# Perform the second fork judgment
print(gini_index(13, 98, 24, 29)) # Chest Pain property
print(gini_index(24, 25, 13, 102)) # Blocked Arteries attribute
# The result is 0.3, 0.29 select the smallest one --- Blocked Arteriers attribute

0.304
0.29


In [ ]:
# Determine whether the third fork is needed
# 24/25 branch
print(gini_index_single(24, 25)) # The result is 0.5
print(gini_index(17, 3, 7, 22)) # The result is 0.321
# So need to fork, and so on